File: neo/backend/src/app/api/v1/knlg_base/rule.py
"""Rule Library API endpoints: Rule CRUD + status + Evidence (READ-ONLY)."""

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.dependencies import get_current_user, get_db
from app.models.user import User
from app.schemas.knlg_base import (
    EvidenceListResponse,
    EvidenceResponse,
    RuleActivateResponse,
    RuleCreate,
    RuleDeprecateResponse,
    RuleListResponse,
    RulePauseResponse,
    RulePublishResponse,
    RuleResponse,
    RuleUpdate,
)
from app.schemas.response import ApiResponse
from app.services.knlg_base.rule import KnlgRuleService

router = APIRouter(prefix="/rules", tags=["knlg-base.rule"])


def get_service(db: Session = Depends(get_db)) -> KnlgRuleService:
    return KnlgRuleService(db)


@router.get("", response_model=ApiResponse[RuleListResponse])
def list_rules(
    workspace_code: str,
    page: int = 1,
    page_size: int = 20,
    source_kc_id: int | None = None,
    status: str | None = None,
    min_confidence: float | None = None,
    keyword: str | None = None,
    service: KnlgRuleService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    items, total = service.list_rules(
        workspace_code,
        current_user,
        page,
        page_size,
        source_kc_id,
        status,
        min_confidence,
        keyword,
    )
    total_pages = (total + page_size - 1) // page_size if total > 0 else 0
    return ApiResponse.success(
        RuleListResponse(
            items=[RuleResponse.model_validate(r) for r in items],
            total=total,
            page=page,
            page_size=page_size,
            total_pages=total_pages,
        )
    )


@router.post("", response_model=ApiResponse[RuleResponse])
def create_rule(
    workspace_code: str,
    data: RuleCreate,
    service: KnlgRuleService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    rule = service.create_rule(workspace_code, current_user, data.model_dump())
    return ApiResponse.success(RuleResponse.model_validate(rule))


@router.get("/{rule_id}", response_model=ApiResponse[RuleResponse])
def get_rule(
    workspace_code: str,
    rule_id: int,
    service: KnlgRuleService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    rule = service.get_rule(workspace_code, current_user, rule_id)
    return ApiResponse.success(RuleResponse.model_validate(rule))


@router.put("/{rule_id}", response_model=ApiResponse[RuleResponse])
def update_rule(
    workspace_code: str,
    rule_id: int,
    data: RuleUpdate,
    service: KnlgRuleService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    rule = service.update_rule(workspace_code, current_user, rule_id, data.model_dump(exclude_unset=True))
    return ApiResponse.success(RuleResponse.model_validate(rule))


@router.delete("/{rule_id}")
def delete_rule(
    workspace_code: str,
    rule_id: int,
    service: KnlgRuleService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    service.delete_rule(workspace_code, current_user, rule_id)
    return ApiResponse.success(None)


@router.post("/{rule_id}/publish", response_model=ApiResponse[RulePublishResponse])
def publish_rule(
    workspace_code: str,
    rule_id: int,
    service: KnlgRuleService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    rule = service.transition_rule(workspace_code, current_user, rule_id, "testing")
    return ApiResponse.success(RulePublishResponse.model_validate(rule))


@router.post("/{rule_id}/activate", response_model=ApiResponse[RuleActivateResponse])
def activate_rule(
    workspace_code: str,
    rule_id: int,
    service: KnlgRuleService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    rule = service.transition_rule(workspace_code, current_user, rule_id, "active")
    return ApiResponse.success(RuleActivateResponse.model_validate(rule))


@router.post("/{rule_id}/pause", response_model=ApiResponse[RulePauseResponse])
def pause_rule(
    workspace_code: str,
    rule_id: int,
    service: KnlgRuleService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    rule = service.transition_rule(workspace_code, current_user, rule_id, "paused")
    return ApiResponse.success(RulePauseResponse.model_validate(rule))


@router.post("/{rule_id}/deprecate", response_model=ApiResponse[RuleDeprecateResponse])
def deprecate_rule(
    workspace_code: str,
    rule_id: int,
    service: KnlgRuleService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    rule = service.transition_rule(workspace_code, current_user, rule_id, "deprecated")
    return ApiResponse.success(RuleDeprecateResponse.model_validate(rule))


# ==================== Evidence (READ-ONLY) ====================


@router.get("/{rule_id}/evidences", response_model=ApiResponse[EvidenceListResponse])
def list_evidences(
    workspace_code: str,
    rule_id: int,
    page: int = 1,
    page_size: int = 20,
    case_source: str | None = None,
    validator_type: str | None = None,
    service: KnlgRuleService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    items, total = service.list_evidences(
        workspace_code,
        current_user,
        rule_id,
        page,
        page_size,
        case_source,
        validator_type,
    )
    total_pages = (total + page_size - 1) // page_size if total > 0 else 0
    return ApiResponse.success(
        EvidenceListResponse(
            items=[EvidenceResponse.model_validate(e) for e in items],
            total=total,
            page=page,
            page_size=page_size,
            total_pages=total_pages,
        )
    )


@router.get("/{rule_id}/evidences/{evidence_id}", response_model=ApiResponse[EvidenceResponse])
def get_evidence(
    workspace_code: str,
    rule_id: int,
    evidence_id: int,
    service: KnlgRuleService = Depends(get_service),
    current_user: User = Depends(get_current_user),
):
    evidence = service.get_evidence(workspace_code, current_user, rule_id, evidence_id)
    return ApiResponse.success(EvidenceResponse.model_validate(evidence))

File: neo/frontend/lib/api/knlg-base/rule.ts
/**
 * Rule Library API client.
 */

import {
  knlgBasePath,
  knlgDelete,
  knlgGet,
  knlgPost,
  knlgPut,
  type Rule,
} from "./_base";

const path = (ws: string, suffix = "") => knlgBasePath(ws, `/rules${suffix}`);

export async function listRules(
  workspaceCode: string,
  params?: {
    page?: number;
    page_size?: number;
    source_kc_id?: number;
    status?: string;
    min_confidence?: number;
    keyword?: string;
  },
): Promise<{
  items: Rule[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}> {
  const search = new URLSearchParams();
  if (params) {
    Object.entries(params).forEach(([k, v]) => {
      if (v !== undefined && v !== null) search.append(k, String(v));
    });
  }
  const qs = search.toString();
  return knlgGet(path(workspaceCode, qs ? `?${qs}` : ""));
}

export async function getRule(
  workspaceCode: string,
  id: number,
): Promise<Rule> {
  return knlgGet(path(workspaceCode, `/${id}`));
}

export async function createRule(
  workspaceCode: string,
  data: {
    name: string;
    description?: string;
    source_kc_id: number;
    scope: Record<string, unknown>;
    trigger: {
      type: string;
      event_name: string;
      filter?: unknown[];
      target_entity?: unknown;
    };
    conditions: Array<{
      field: string;
      operator: string;
      value: unknown;
      combinator?: string;
    }>;
    conclusion: Record<string, unknown>;
    exceptions?: Array<Record<string, unknown>>;
    confidence?: number;
  },
): Promise<Rule> {
  return knlgPost(path(workspaceCode), data);
}

export async function updateRule(
  workspaceCode: string,
  id: number,
  data: Partial<{
    name: string;
    description: string;
    trigger: unknown;
    conditions: unknown[];
    conclusion: unknown;
    exceptions: unknown[];
    confidence: number;
  }>,
): Promise<Rule> {
  return knlgPut(path(workspaceCode, `/${id}`), data);
}

export async function deleteRule(
  workspaceCode: string,
  id: number,
): Promise<null> {
  return knlgDelete(path(workspaceCode, `/${id}`));
}

export async function publishRule(
  workspaceCode: string,
  id: number,
): Promise<Rule> {
  return knlgPost(path(workspaceCode, `/${id}/publish`), {});
}

export async function activateRule(
  workspaceCode: string,
  id: number,
): Promise<Rule> {
  return knlgPost(path(workspaceCode, `/${id}/activate`), {});
}

export async function pauseRule(
  workspaceCode: string,
  id: number,
): Promise<Rule> {
  return knlgPost(path(workspaceCode, `/${id}/pause`), {});
}

export async function deprecateRule(
  workspaceCode: string,
  id: number,
): Promise<Rule> {
  return knlgPost(path(workspaceCode, `/${id}/deprecate`), {});
}

export async function listEvidences(
  workspaceCode: string,
  ruleId: number,
  params?: {
    page?: number;
    page_size?: number;
    case_source?: string;
    validator_type?: string;
  },
): Promise<{ items: unknown[]; total: number }> {
  const search = new URLSearchParams();
  if (params) {
    Object.entries(params).forEach(([k, v]) => {
      if (v !== undefined && v !== null) search.append(k, String(v));
    });
  }
  const qs = search.toString();
  return knlgGet(
    path(workspaceCode, `/${ruleId}/evidences${qs ? `?${qs}` : ""}`),
  );
}
